tests/unit/s2n_certificate_test.c

```diff
@@ -754,7 +755,184 @@ int main(int argc, char **argv)
+            EXPECT_TRUE(s2n_connection_check_io_status(server, S2N_IO_CLOSED));
+            EXPECT_FALSE(s2n_connection_check_io_status(client, S2N_IO_CLOSED));
+
+            EXPECT_NOT_EQUAL(server->handshake_params.client_cert_chain.size, 0);
+            EXPECT_NOT_NULL(server->handshake_params.client_cert_chain.data);
+        };
+
+        /* Test: Malformed TLS1.2 certificate chain is still available */
+        {
+            DEFER_CLEANUP(struct s2n_stuffer input, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&input, 0));
+
+            struct s2n_stuffer_reservation total_size = { 0 };
+            const uint32_t cert_size = 24;
+            EXPECT_SUCCESS(s2n_stuffer_reserve_uint24(&input, &total_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint24(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_skip_write(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint24(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_skip_write(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_vector_size(&total_size));
+
+            /* Sanity check: test cert chain is well-formed (but invalid) */
+            {
+                DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                        s2n_connection_ptr_free);
+                server->actual_protocol_version = S2N_TLS12;
+                EXPECT_SUCCESS(s2n_stuffer_copy(&input, &server->handshake.io,
+                        s2n_stuffer_data_available(&input)));
+
+                EXPECT_FAILURE_WITH_ERRNO(s2n_client_cert_recv(server), S2N_ERR_CERT_INVALID);
+                EXPECT_NOT_EQUAL(server->handshake_params.client_cert_chain.size, 0);
+                EXPECT_NOT_NULL(server->handshake_params.client_cert_chain.data);
+            };
+
+            /* Modify last cert to be malformed. Keep correct total size */
+            EXPECT_SUCCESS(s2n_stuffer_reread(&input));
+            EXPECT_SUCCESS(s2n_stuffer_wipe_n(&input, 1));
+            EXPECT_SUCCESS(s2n_stuffer_write_vector_size(&total_size));
+
+            /* Data available even if malformed: we don't parse pre-TLS1.3 chains
+             * before storing them.
+             */
+            {
+                DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                        s2n_connection_ptr_free);
+                server->actual_protocol_version = S2N_TLS12;
+                EXPECT_SUCCESS(s2n_stuffer_copy(&input, &server->handshake.io,
+                        s2n_stuffer_data_available(&input)));
+
+                EXPECT_FAILURE_WITH_ERRNO(s2n_client_cert_recv(server), S2N_ERR_CERT_INVALID);
+                EXPECT_NOT_EQUAL(server->handshake_params.client_cert_chain.size, 0);
+                EXPECT_NOT_NULL(server->handshake_params.client_cert_chain.data);
+            };
+        };
+
+        /* Test: Malformed TLS1.3 certificate chain is not available */
```

Comment:
Isn't this test showing that the malformed TLS1.3 cert chain _is_ available?

Reply:
No. See line 923-924. Which lines do you think show otherwise?

Reply:
I think this goes back to my above comment, where your sanity check looks like a real test to me.